### DISCUSSION TITLE
custom plugin events

### DISCUSSION THREAD

**Discussion by 3l3ktr0n** [e107inc/e107#4828] [Q&A Plugins] [answered]
Is there a way to create custom event triggers?

**Comment by Jimmi08** [comment]
You can check e_event.php in some plugins.  If not core ones,  lonalore plugins are good examples too.

**Reply by 3l3ktr0n** [reply]
Isn't what I'm really looking for. All examples are with core event triggers and own function hooks. What I want is to create custom triggers.

**Reply by 3l3ktr0n** [reply]
Or am I misunderstanding something?

**Reply by Jimmi08** [reply]
I would look at handlers/event_class.php

**Register:**
`https://e107.org/developer-manual/classes-and-methods#events`
says
`e107::getEvent()->register(name, function, include);`


**Trigger** - see description in gsitemap/e_event.php
- 
```
* Developers can trigger their own events using: e107::getEvent()->trigger('plugin_event',$array);
* Where 'plugin' is the folder of their plugin and 'event' is a unique name of the event. * $array is data which is sent to the triggered function. eg. myfunction($array) in the example below.
```

Sorry, I can't remember more.

**Reply by 3l3ktr0n** [reply]
to my knowledge this doesn't help
register is to hook your function to a trigger.. like when a user logins (that is a core trigger) myFunction will be run
what I need is to know how to make custom triggers and where are the core ones defined

```
e107::getEvent()->register('login', 'myFunction'); 

function myFunction($data)
{
   // do something    
}
```
like when some event happens that isnt a core trigger I can run my function

**Reply by Jimmi08** [reply]
See plugin 
https://github.com/lonalore/nodejs
plugin event is registered in e_module

**Reply by 3l3ktr0n** [reply]
> See plugin https://github.com/lonalore/nodejs plugin event is registered in e_module

it uses core triggers

**Reply by Jimmi08** [reply]
See this issue, there was example from that plugin, it was probably changed

https://github.com/e107inc/e107/issues/880

**Comment by Jimmi08** [comment]
I just needed this, so while I am on it.

**Plugin with the event:**

- in e_module.php
```
if(e_ADMIN_AREA === true)  
{
}
else {
	$event = e107::getEvent();
	$event->register('survey', 'displayed');
}

 
function displayed($data = array()) {
}
```

trigger in survey.php (in the same plugin)

`$JobAvail = e107::getEvent()->trigger('survey_displayed', $array);`

and using this event with other plugin
in e_event.php

```
class consultant_event
{

	function config()
	{
		$survey = array();

		$survey[] = array(
			'name'	=> "survey_displayed", 
			'function'	=> "consultant_profile"
		);

		return $survey;

	}
	function consultant_profile($data)
	{
	      //code
	}
}
```
it works.

**Reply by 3l3ktr0n** [reply]
it works?

I'm wondering how and when is `$event->register('survey', 'displayed');` is survey getting triggered?
When  is exactly that 'survey' happening that you're hooking into?

`$JobAvail = e107::getEvent()->trigger('survey_displayed', $array);` how do they hook into each other?

I'm somewhat confused. @CaMer0n @Moc do you have any insight on the e_events matter?

**Reply by Moc** [reply]
@3l3ktr0n I've posted up an answer.

**Reply by Moc** [reply]
@Jimmi08 I've removed my previous answer and your comments to avoid confusion. See my new answer.

**Comment by Moc** [comment, marked as answer]
Yes.  Simply call the event anywhere in your code by using  `e107::getEvent()->trigger('myplugin_customevent', $data)`. 

- `myplugin` is your plugin folder name
- `customevent` is a unique event name

You do not necessarily need to add your custom plugin event to YOUR  `e_event` addon. If you trigger it somewhere in your code, other plugins can "listen" to your custom event by defining it in THEIR `e_event` addon. 

------------
**OPTIONALLY:**
If you want to LISTEN to your OWN custom event and run a function when it is triggered, then also add the custom event your `e_event.php` addon, within the `config()` method. Also add the method that you want to run on that event. 

For example:
```php
function config()
{
	$event = array();

	$event[] = array(
		'name'		=> "myplugin_customevent", // "myplugin" is the plugin folder name. "customevent" should be a unique event name. 
		'function'	=> "customfunction",
	);

	return $event;
}

function customfunction($data)
{
	echo "Do something!"; 
}
```


**Notes:**
- If you create the e_event addon after installing your plugin, you may need to run a "Scan plugin directories" in `Admin Area > Tools > Database` for e107 to pick it up (and register the event(s)).

- I've added new examples to the [`e_event` addon of the `_blank` plugin](https://github.com/e107inc/e107/blob/master/e107_plugins/_blank/e_event.php): 
(And I'll also add it to the new developer guide). 

- Previously (before e107 v2.1.2), one had to use the `e_module` addon and `e107::getEvent()->register()`. You may still find examples of older plugins that use the `e_module` addon. This is not recommended, please use the `e_event` addon instead.
